File: maximizeSAAROCF.js
/**
 * @param {number} n
 * @param {number[][]} conflictingPairs
 * @return {number}
 */
var maxSubarrays = function(n, conflictingPairs) {
    // conflicts[r] stores left endpoints that conflict with subarrays ending at r
    const conflicts = Array.from({ length: n + 1 }, () => []);

    for (const [a, b] of conflictingPairs) {
        const maxVal = Math.max(a, b);
        const minVal = Math.min(a, b);
        conflicts[maxVal].push(minVal);
    }

    let totalValidSubarrays = 0;
    const gains = Array(n + 1).fill(0);
    let maxLeft = 0;
    let secondMaxLeft = 0;

    for (let right = 1; right <= n; right++) {
        for (const left of conflicts[right]) {
            if (left > maxLeft) {
                secondMaxLeft = maxLeft;
                maxLeft = left;
            } else if (left > secondMaxLeft) {
                secondMaxLeft = left;
            }
        }

        totalValidSubarrays += right - maxLeft;

        if (maxLeft > 0) {
            gains[maxLeft] += maxLeft - secondMaxLeft;
        }
    }

    const maxGain = Math.max(...gains);
    return totalValidSubarrays + maxGain;
};

File: maximizeSAAROCF.java
import java.util.*;

class Solution {
    public long maxSubarrays(int n, int[][] conflictingPairs) {
        // conflicts[r] stores the left endpoints that conflict with subarrays ending at r
        List<List<Integer>> conflicts = new ArrayList<>();
        for (int i = 0; i <= n; i++) {
            conflicts.add(new ArrayList<>());
        }

        for (int[] pair : conflictingPairs) {
            int a = pair[0];
            int b = pair[1];
            conflicts.get(Math.max(a, b)).add(Math.min(a, b));
        }

        long totalValidSubarrays = 0;
        long[] gains = new long[n + 1];
        int maxLeft = 0;
        int secondMaxLeft = 0;

        for (int right = 1; right <= n; right++) {
            for (int left : conflicts.get(right)) {
                if (left > maxLeft) {
                    secondMaxLeft = maxLeft;
                    maxLeft = left;
                } else if (left > secondMaxLeft) {
                    secondMaxLeft = left;
                }
            }

            // Subarrays from [maxLeft + 1, right] to [right, right] are valid
            totalValidSubarrays += right - maxLeft;

            // If we remove maxLeft, we gain subarrays from [secondMaxLeft + 1, right] to [maxLeft, right]
            if (maxLeft > 0) {
                gains[maxLeft] += maxLeft - secondMaxLeft;
            }
        }

        long maxGain = 0;
        for (long gain : gains) {
            maxGain = Math.max(maxGain, gain);
        }

        return totalValidSubarrays + maxGain;
    }
}
